spam/03_spam_data_slicing_tutorial.py

```diff
@@ -0,0 +1,483 @@
+    return len(x.doc) < 20 and any([ent.label_ == "PERSON" for ent in x.doc.ents])
+
+
+from textblob import TextBlob
+
+
+@preprocessor(memoize=True)
+def textblob_sentiment(x):
+    scores = TextBlob(x.text)
+    x.polarity = scores.sentiment.polarity
+    x.subjectivity = scores.sentiment.subjectivity
+    return x
+
+
+@slicing_function(pre=[textblob_sentiment])
+def textblob_polarity(x):
+    return x.polarity > 0.9
+
+
+extra_sfs = [
+    keyword_subscribe,
+    keyword_please,
+    regex_check_out,
+    short_comment,
+    has_person_nlp,
+    textblob_polarity,
+]
+
+sfs = [short_link] + extra_sfs
+
+# %%
+applier = PandasSFApplier(sfs)
+S_train = applier.apply(df_train)
+S_valid = applier.apply(df_valid)
+S_test = applier.apply(df_test)
+
+# %%
+slice_names = [sf.name for sf in sfs]
+add_slice_labels(dl_train, spam_task, S_train, slice_names)
+add_slice_labels(dl_valid, spam_task, S_valid, slice_names)
+add_slice_labels(dl_test, spam_task, S_test, slice_names)
+
+# %% [markdown]
+# Like we saw above, we'd like to visualize examples in the slice.
+
+# %%
+pd_slicer = PandasSlicer(df_valid)
+polarity_df = pd_slicer.slice(textblob_polarity)
+polarity_df[["text", "label"]].head()
+
+# %% [markdown]
+# ### Representation learning with slices
+
+# %% [markdown]
+# To cope with scale, we will attempt to learn and combine many slice-specific representations with an attention mechanism (for more, please see our technical report — coming soon!).
+# Using the helper, `convert_to_slice_tasks`, we have now have a list of slice tasks that appropriate constructs the `task_flow` to do just that!
+
+# %%
+from snorkel.slicing.utils import convert_to_slice_tasks
+
+slice_tasks = convert_to_slice_tasks(spam_task, slice_names)
+slice_tasks
+
+# %%
+slice_model = SnorkelClassifier(slice_tasks)
+
+# %% [markdown]
+# We train this model, and note that we can monitor slice-specific performance during training.
+# This is a powerful way to track especially critical subsets of the data.
+#
+# _Note: This model includes more parameters (corresponding to additional slices) — we only train for 1 epoch here for demonstration purposes._
+
+# %%
+trainer = Trainer(n_epochs=1, lr=1e-4, progress_bar=True)
+trainer.fit(slice_model, [dl_train, dl_valid])
+
+# %% [markdown]
+# At inference time, the primary task head (`spam_task`) will be making all final predictions.
+# We'd like to evaluate all the slice heads on the original task head.
+# To do this, we use our `remap_labels` API, as we did earlier.
+# Note that this time, we map each `ind` head to `None` — it doesn't make sense to evaluate these labels on the base task head.
+
+# %%
+Y_dict = dl_valid.dataset.Y_dict
+eval_mapping = {label: "spam_task" for label in Y_dict.keys() if "pred" in label}
+eval_mapping.update({label: None for label in Y_dict.keys() if "ind" in label})
+
+# %% [markdown]
+# _Note: in this toy dataset, we might not see significant gains because slices are defined for demo purposes. 
```

Comment:
because our dataset is so small that (i) there are few examples the train split, giving little signal to learn over, and (ii) there are few examples int he test split, making our evaluation metrics very noisy.